experiment/secret-sync-controller/main.go

```diff
@@ -1,110 +1,199 @@
-	"log"
 	"os"
-	"reflect"
+	"time"
 )
 
 type options struct {
-	configPath string
+	configPath   string
+	testSetup    string
+	runOnce      bool
+	resyncPeriod int64
 }
 
 func (o *options) Validate() error {
 	if o.configPath == "" {
 		return errors.New("required flag --config-path was unset")
 	}
-
 	return nil
 }
 
 func gatherOptions() options {
 	o := options{}
 	flag.StringVar(&o.configPath, "config-path", "", "Path to config.yaml.")
+	flag.StringVar(&o.testSetup, "test-setup", "", "Path to test-setup.yaml.")
+	flag.BoolVar(&o.runOnce, "run-once", false, "Sync once instead of continuous loop.")
+	flag.Int64Var(&o.resyncPeriod, "period", 1000, "Resync period in milliseconds.")
 	flag.Parse()
 	return o
 }
 
+type SecretSyncController struct {
+	Client       client.Interface
+	Config       *SecretSyncConfig
+	RunOnce      bool
+	ResyncPeriod time.Duration
+	// TODO: in-struct looger?
+}
+
 func main() {
 	o := gatherOptions()
 
-	// TODO: modularize clients
+	// prepare clients
 	k8sClientset, err := client.NewK8sClientset()
 	if err != nil {
-		log.Fatal(err)
+		klog.Errorf("New kubernetes client failed: %s", err)
 	}
 	secretManagerClient, err := client.NewSecretManagerClient(context.Background())
 	if err != nil {
-		log.Fatal(err)
+		klog.Errorf("New Secret Manager client failed: %s", err)
+	}
+	clientInterface := &client.Client{
+		K8sClientset:        *k8sClientset,
+		SecretManagerClient: *secretManagerClient,
 	}
 
-	var clientInterface ClientInterface
-
-	client := Client{
-		K8sClientset:        k8sClientset,
-		SecretManagerClient: secretManagerClient,
+	// prepare config
+	secretSyncConfig, err := LoadConfig(o.configPath)
+	if err != nil {
+		klog.Errorf("Load config failed: %s", err)
 	}
 
-	clientInterface = client
+	controller := &SecretSyncController{
+		Client:       clientInterface,
+		Config:       secretSyncConfig,
+		RunOnce:      o.runOnce,
+		ResyncPeriod: time.Duration(o.resyncPeriod) * time.Millisecond,
+	}
 
-	secretSyncConfig, err := LoadConfig(o.configPath)
+	err = controller.ValidateAccess()
 	if err != nil {
-		log.Fatal(err)
+		klog.Error(err)
 	}
 
-	for i, spec := range secretSyncConfig.Specs {
-		fmt.Printf("Secret pair [%d]:\n{\n", i)
-		err := Sync(clientInterface, spec)
-		if err != nil {
-			log.Fatal(err)
+	var stopChan <-chan struct{}
+	controller.Start(stopChan)
+
+}
+
+func (c *SecretSyncController) Start(stopChan <-chan struct{}) error {
+	runChan := make(chan struct{})
+
+	go func() {
+		for {
+			runChan <- struct{}{}
+			time.Sleep(c.ResyncPeriod)
+		}
+	}()
+
+	for {
+		select {
+		case <-stopChan:
+			klog.Info("Stop signal received. Quitting...")
+			return nil
+		case <-runChan:
+			c.SyncAll()
+			if c.RunOnce {
+				c.Client.CleanupKubernetesSecrets("ns1")
+				return nil
+			}
+		}
+	}
+}
+
+// ValidateAccess validates the required access for each sync pair.
+// If finds an invlaid access, return an error and delete that sync pair.
```

Comment:
That makes a lot of sense.
Now this will only be tackled in `SyncAll()`. Any pair of secrets that fails to be synced/accessed will be skipped and an error message for that secret pair will pop up.